RavenDB-19992: Initial implementation of the CompactKey abstraction

redknightlois opened the pull request:
Improving performance by diminishing the amount of encode/decode cycles required to query and index. 

Issues:
https://issues.hibernatingrhinos.com/issue/RavenDB-20076
https://issues.hibernatingrhinos.com/issue/RavenDB-19992 [partial]

Non breaking changes.

redknightlois committed:
RavenDB-19992: Initial implementation of the CompactKey abstraction

ayende commented on src/Corax/IndexSearcher/IndexSearcher.TermMatch.cs:
```diff
@@ -63,6 +63,11 @@ internal TermMatch TermQuery(in FieldMetadata field, CompactTree tree, Slice ter
     
+    public TermMatch TermQuery(FieldMetadata field, CompactKey term, CompactTree termsTree = null)
+    {
+        return TermQuery(field, termsTree, term.Decoded());
```
do we have to decode it?

ayende commented on src/Corax/Queries/TermProviders/TermProvider.Exists.cs:
```diff
@@ -43,16 +46,18 @@ public bool Next(out TermMatch term)
             {
-                int termSize = termSlice.Length;
-                if (termSlice.Length > 1)
+                var key = keyScope.Key.Decoded();
```
Why do we need to decode for `Exists` ? 
That is just because we expose the term for the `TermMatch` ?

redknightlois replied:
Because for now `TermMatch` doesnt use the key abstraction.

ayende commented on src/Corax/Queries/TermProviders/TermProvider.NotStartsWith.cs:
```diff
@@ -33,10 +35,10 @@ public void Reset()
             {
-                if (termSlice.StartWith(_startWith))
+                var termSlice = termScope.Key.Decoded();
+                if (termSlice.StartsWith(_startWith))
```
Can't we do starts with on the encoded value?

redknightlois replied:
Pretty sure we can find a way to do it... Yes... haven't focused on that yet.

ayende commented on src/Corax/Queries/TermProviders/TermProvider.Range.cs:
```diff
@@ -65,29 +67,31 @@ public void Reset()
                 return false;
             }
 
+            var termSlice = termScope.Key.Decoded();
```
Can we do comparision on the enconded value?

ayende commented on src/Corax/Queries/TermProviders/TermProvider.StartWith.cs:
```diff
@@ -14,36 +14,50 @@ namespace Corax.Queries
+            }
+
+            var key = termScope.Key.Decoded();
+            if (key.StartsWith(_startWith) == false)
```
Can't we do this on the encoded value?

redknightlois replied:
Most likely we can, I still have to figure out how.

ayende commented on src/Voron/Data/CompactTrees/CompactKey.cs:
```diff
@@ -0,0 +1,432 @@
+                {
+                    Avx.Store(_keyMappingCache + currentElementIdx, Avx.LoadDquVector256(key._keyMappingCache + currentElementIdx));
+                    Avx.Store(_keyMappingCacheIndex + currentElementIdx, Avx.LoadDquVector256(key._keyMappingCacheIndex + currentElementIdx));
+                    currentElementIdx += Vector256<long>.Count;
```
Isn't this possible to copy _more_ than the buffer size?

redknightlois replied:
Good catch!!! I cannot think how would you end up in that situation 'realistically' but we need to handle that case either way.

ayende commented on src/Voron/Data/CompactTrees/CompactKey.cs:
```diff
@@ -0,0 +1,432 @@
+
+namespace Voron.Data.CompactTrees;
+
+public unsafe class CompactKey : IDisposable
```
Please explain here intended use case

ayende replied:
There is _one_ key at a given time?

redknightlois committed:
RavenDB-19992: Fixed a potential invalid memory access

redknightlois committed:
RavenDB-19992: Implement upgrade on insertion of dictionaries

This is interesting because it had the aftereffect of making indexing even faster.

ayende commented on src/Voron/Data/CompactTrees/CompactTree.cs:
```diff
@@ -1897,52 +1448,40 @@ private void FindPageFor(EncodedKey key, ref IteratorCursorState cstate)
+                // we can offset the cost of future key re-encoding along the way. We could choose to do this either before or after
+                // adding/searching, but performing it proactively guarantees that we avoid the need for re-encoding at this stage.
+                if (tryRecompress && _state.TreeDictionaryId != state.Header->DictionaryId)
+                    TryRecompressPage(ref state);
```
This deals with all the branch pages, and will proactively move them
Given that we aren't deleting dictionaries, what is the current trigger for generating a new one?

ayende replied:
Also, you are updating the _branches_, but I think that the only time we'll recompress a leaf page is when we are out of space, no?

https://github.com/ravendb/ravendb/blob/c62101b8381fbe33c0825aa9997c76f8cb3b475e/src/Voron/Data/CompactTrees/CompactTree.cs#L958

If we are proactive, shouldn't we always do that when we modify a leaf page? (Note, we have to deal with the newly compressed page being _too big_ to fit the page in that scenario.

redknightlois replied:
Mhhh, it would make sense to do that proactively too for leaf, even though it is going to happen when splits happen (more often). 

When the recompress page fails, it just keep retrying on write until it is able to do so. We could add a marker for that, but it is not such a big deal; considering that the fallback is just reencode the key for that dictionary anyways... We are playing the game that eventually it will happen.

redknightlois replied:
The generation of new dictionaries follows a formula that makes it more difficult to trigger the bigger the tree is.

redknightlois commented:
Not necessarily, there can be as many `CompactKey` as our ability to reuse them. For example in the indexing process we have many because comparisons happen on multiple trees over time and would carry over all the different potential representations along the indexing process. The more we reuse the better the gain.

redknightlois committed:
RavenDB-20076: It is found safe to just ignore the prefix-less key.

redknightlois committed:
RavenDB-19992: Improve access locality.

redknightlois committed:
RavenDB-19992: Global Shared Persistent Dictionaries.

redknightlois commented:
Now the main includes Global Shared Persistent Dictionaries

redknightlois committed:
RavenDB-19992: Update primitives to use Compact Keys.
